### PR TITLE
Optimizing GPU L2 regularization term.

### PR DESCRIPTION
**Pull request by timmoon10**
I am attempting to reduce the time spent in forward prop for the L2 regularization term on GPU. Changes:
- Using the cuBLAS `dot` routine instead of `nrm2`, which uses a numerically stable algorithm that is likely excessive for our use case.
- Running cuBLAS kernels asynchronously w.r.t. the host.
- Using all available GPUs.

### CODE REVIEW

File: include/lbann/objective_functions/weight_regularization/l2.hpp [hunk 1]
```diff
@@ -64,12 +64,17 @@ class l2_weight_regularization : public objective_function_term {
   void compute_weight_regularization() override;
 
  private:
+
+  /** Reference to cuDNN manager. */
+  cudnn::cudnn_manager* m_cudnn = nullptr;
```
Comment (bvanessen) [inline include/lbann/objective_functions/weight_regularization/l2.hpp]: Shouldn't this be wrapped in #ifdef LBANN_HAS_CUDNN

File: src/objective_functions/weight_regularization/l2.cpp [hunk 2]
```diff
@@ -88,51 +85,83 @@ void l2_weight_regularization::setup(model& m) {
   for (size_t i = 0; i < m_weights.size(); ++i) {
     m_allreduce_reqs.push_back(std::move(Al::request()));
   }
+
+  // Get cuDNN manager
+  m_cudnn = nullptr;
```
Comment (bvanessen) [inline src/objective_functions/weight_regularization/l2.cpp]: Shouldn't this be wrapped in #ifdef LBANN_HAS_CUDNN

**Comment by ndryden** [comment]
It appears that L2 regularization, in particular, remains the primary component of forward and backward prop runtime for the objective function. We should look at optimizing this further.

**Comment by ndryden** [comment]
This looks good now, :shipit: .

**Pull request merged**